beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -357,22 +358,31 @@ func (f *blocksFetcher) requestBlocks(
 		}
 	}()
 
-	resp := make([]*eth.SignedBeaconBlock, 0, req.Count)
+	blocks := make([]*eth.SignedBeaconBlock, 0, req.Count)
 	for i := uint64(0); ; i++ {
 		isFirstChunk := i == 0
 		blk, err := prysmsync.ReadChunkedBlock(stream, f.p2p, isFirstChunk)
 		if err == io.EOF {
 			break
 		}
-		// exit if more than max request blocks are returned
+		// Exit if more than max request blocks are returned.
 		if i >= params.BeaconNetworkConfig().MaxRequestBlocks {
 			break
 		}
 		if err != nil {
 			return nil, err
 		}
-		resp = append(resp, blk)
+		// Returned blocks MUST be in the slot range [start_slot, start_slot + count * step).
+		if blk.Block.Slot < req.StartSlot || blk.Block.Slot >= req.StartSlot+req.Count*req.Step {
+			return nil, errInvalidFetchedData
+		}
+		blocks = append(blocks, blk)
+	}
+
+	// The response MUST contain no more than `count` blocks.
+	if uint64(len(blocks)) > req.Count {
```

Comment:
maybe bring this check up in the loop ? Otherwise a peer could make us read the whole batch of 1024 blocks instead of like the 10 we requested.

Reply:
Thank you, nice catch - will do.